rockcraft/pebble.py

```diff
@@ -16,81 +16,168 @@
+    check spec in the plan with the same name.
+    """
     level: Literal["alive", "ready"] | None = None
-    period: str | None = None
-    timeout: str | None = None
-    threshold: int | None = None
-    http: HttpCheck | None = None
-    tcp: TcpCheck | None = None
-    exec: ExecCheck | None = None
-
-    @pydantic.model_validator(mode="before")
-    @classmethod
-    def _validates_check_type(cls, values: Mapping[str, Any]) -> Mapping[str, Any]:
-        """Before validation, make sure only one of 'http', 'tcp' or 'exec' exist."""
-        mutually_exclusive = ["http", "tcp", "exec"]
-        check_types = list(set(mutually_exclusive) & values.keys())
-        check_types.sort()
-
-        if len(check_types) > 1:
-            err = str(
-                f"Multiple check types specified ({', '.join(check_types)}). "
-                "Each check must have exactly one type."
+    """Check level, used for filtering checks when calling the health endpoint.
+
+    ``ready`` implies ``alive``, and not ``alive`` implies not ``ready``,
+    but not the other way around.
+    More information is available in the pebble documentation:
+    :external+pebble:doc:`reference/health-checks`
+    """
+    period: str | None = pydantic.Field(
+        default=None,
+        description="How frequently to run this check.",
+        examples=["10s", "30m"],
+    )
+    timeout: str | None = pydantic.Field(
+        default=None,
+        description="The time before the check fails if not completed. Must be less than ``period``.",
+        examples=["5s", "2m"],
+    )
+    threshold: Annotated[int, annotated_types.Ge(1)] | None = pydantic.Field(
+        default=None,
+        description="Number of consecutive errors before the check is considered failed.",
+    )
+
+
+class HttpCheck(_BaseCheck):
+    """Model for a check that uses HTTP."""
+
+    http: HttpCheckOptions
+
+
+class TcpCheck(_BaseCheck):
+    """Model for a check that uses TCP."""
+
+    tcp: TcpCheckOptions
+
+
+class ExecCheck(_BaseCheck):
+    """Model for a check that executes a command."""
+
+    exec: ExecCheckOptions
+
+
+def _get_check_tag(check: Mapping[str, Any]) -> str:
+    tags = ("http", "tcp", "exec")
+    check_types = check.keys() & tags
+    match len(check_types):
+        case 0:
+            raise CraftValidationError(
+                f"Must specify exactly one of {', '.join(tags)} for each check."
             )
-        elif len(check_types) < 1:
-            err = str(
-                f"Must specify exactly one of {', '.join(list(mutually_exclusive))} for each check."
+        case 1:
+            return check_types.pop()
+        case _:
+            raise CraftValidationError(
+                f"Multiple check types specified ({', '.join(sorted(check_types))}). "
+                "Each check must have exactly one type."
             )
-        else:
-            return values
 
-        raise CraftValidationError(err)
+
+Check = Annotated[
```

Comment:
This change means that the schema is now aware that these fields are mutually exclusive. It's a tagged union, so we still get a nice error message out of it. No new tests are needed, as `_get_check_tag` is tested with the same tests that previously tested `_validates_check_type`. I did add some test cases though.